gourmet/timer.py

```diff
@@ -1,99 +1,128 @@
-from gi.repository import Gtk
-from gi.repository import GObject
-import time
-from . import gglobals
 import os
+import time
+from typing import Callable, Optional
+
+from gettext import gettext as _
+from gi.repository import Gtk, GObject
 import xml.sax.saxutils
+
+from . import gglobals
 from .sound import Player
 from .gtk_extras import cb_extras as cb
-from gettext import gettext as _
+from .gtk_extras.dialog_extras import getBoolean, UserCancelledError
+
 
 class TimeSpinnerUI:
 
-    def __init__ (self, hoursSpin, minutesSpin, secondsSpin):
-        self.timer_hooks = []
-        self.running = False
+    def __init__(self,
+                 hoursSpin: Gtk.SpinButton,
+                 minutesSpin: Gtk.SpinButton,
+                 secondsSpin: Gtk.SpinButton):
+        self.timer_hooks = []  # Actions to run when timer is over
+        self.is_running = False  # State flag
+        self.elapsed = 0  # The time already spent
+        self.previous_iter_time = None  # epoch stamp, used for loop ticks
+        self.orig_time = 0  # in seconds, the user input time, used for reset
```

Comment:
since we are here, can you type-annotate the instance properties as well?

Reply:
done :)